### PR TITLE
Fixed key-shared delivery of messages with interleaved delays

### LINKED ISSUE #15397: Delayed message delivery intermittently fails to deliver messages until consumers reconnect when a KeyShared consumer is used and delays are randomized
**Describe the bug**
When consuming from a topic with a KeyShared consumer, messages with delayed delivery are often not returned from the consumer. This doesn't happen with all delayed messages, but seems to happen with around half.

This issue doesn't occur when using Shared subscriptions.

After the consumer is restarted, all messages that should have been delivered but were not are immediately delivered.

This only happens when delayed messages are not sequentially scheduled; i.e. it won't happen if I publish messages in order with equal or increasing delays, since those messages "come due" in publish order. For example, it will happen if I publish two messages, the first with a delay of 10sec and the second with a delay of 5sec.

**To Reproduce**
1. In one terminal, run the below code snippet with the command line argument `consume`. Verify it connects and waits for messages.
2. In another terminal, run the below snippet with the command line argument `produce`; observe that it prints out messages that it published and their delays.
3. In the first terminal, wait for 12 seconds (the maximum possible delay published by the producer script). 
4. In the first terminal, observe that all messages were *not* delivered.
5. Restart (CTRL+C and run again) the consumer in the first terminal. Observe that it immediately receives all of the messages that were missed in step 4.

Code to reproduce:
```python
import sys
import time
from datetime import timedelta, datetime
import random
import pulsar

topic = "persistent://public/default/foobar"


def main():
    client = pulsar.Client(service_url='pulsar://localhost:6650')
    if sys.argv[-1] == "produce":
        p = client.create_producer(topic=topic)
        for i in range(10):
            data = f"msg-{i}".encode()
            delay = timedelta(milliseconds=random.randint(1000, 2000))
            rv = p.send(data, deliver_after=delay, partition_key=str(i))
            print("sent", data, "delay:", delay, rv)
            time.sleep(1)
    if sys.argv[-1] == "consume":
        c = client.subscribe(
            topic=topic,
            subscription_name="chariot_subscription-heartbeat-heartbeat",
            consumer_type=pulsar.ConsumerType.KeyShared
        )

        while True:
            try:
                msg = c.receive(timeout_millis=1000)
            except pulsar.Timeout:
                pass
            else:
                c.acknowledge(msg)
                print(datetime.now(), msg.message_id(), msg.data())


if __name__ == "__main__":
    main()
```

**Expected behavior**
All messages should print out as received in step 3 above.

**Desktop (please complete the following information):**
 - OS: MacOS monterey.
 - Pulsar: 2.9.2 via Homebrew; also repro'd on latest standalone Docker image as of 4/29.

**Additional context**

KeyShared consumers are necessary for our use case: topics sometimes contain delayed messages, and sometimes contain **non**-delayed messages with a partition key set across groups of messages for ordered, single-consumer delivery. 

We know that combining delays with a partition key can cause problems; so we enforce that delayed messages have UUID partition keys at our client layer. 

To facilitate delayed delivery, we could work around the problem by periodically reconnecting consumers (hacky, induces extra delays) or by running a second Shared subscriber next to our KeyShared subscribers (complicated and potentially expensive in terms of consumer memory). We would prefer to keep the system simple and use a single KeyShared consumer for both types of messages.

### PR DESCRIPTION
**Pull request by merlimat**
### Motivation

Fixes #15397

The changes in #12890 introduced a regression that when delayed messages with interleaved delays occur on a shared subscriptions, many of the messages are not delivered and stay in the backlog. 

The reason is that we're peeking into `getMessagesToReplayNow()` and we cannot discard the returned set, because these message ids are not tracked anymore in the delayed messages controller.

### CODE REVIEW

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherMultipleConsumers.java [hunk 1]
```diff
@@ -270,7 +270,11 @@ public synchronized void readMoreEntries() {
                             consumerList.size());
                 }
                 havePendingRead = true;
-                minReplayedPosition = getMessagesToReplayNow(1).stream().findFirst().orElse(null);
+                Set<PositionImpl> toReplay = getMessagesToReplayNow(1);
+                minReplayedPosition = toReplay.stream().findFirst().orElse(null);
```
Comment (eolivelli) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherMultipleConsumers.java:274]: Why 'findFirst'? It is an unordered set

**Review by eolivelli — CHANGES_REQUESTED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherMultipleConsumers.java (hunk 1 above)
Comment (merlimat) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherMultipleConsumers.java:274]: It’s a set with either 0 or 1 entry

**Review by eolivelli — APPROVED** [review]
Lgtm

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherMultipleConsumers.java (hunk 1 above)
Comment (eolivelli) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherMultipleConsumers.java:274]: Oh, got it.
It wasn't clear from the type.
Tha ks for your clarification

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentStickyKeyDispatcherMultipleConsumers.java [hunk 2]
```diff
@@ -174,29 +174,36 @@ protected void sendMessagesToConsumers(ReadType readType, List<Entry> entries) {
...
+                            });
+                        } else if (readType == ReadType.Replay) {
+                            entries.forEach(Entry::release);
+                        }
+                        readMoreEntries();
+                        return;
                     }
-                    readMoreEntries();
-                    return;
+                } else {
```
Comment (Jason918) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentStickyKeyDispatcherMultipleConsumers.java:202]: Why only do this in `else` part？
It seems `replayPosition` could be popped from `delayedDeliveryTracker` if `minReplayedPosition` is null.

Comment (merlimat) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentStickyKeyDispatcherMultipleConsumers.java:202]: @Jason918 You are absolutely correct, we need to do in both cases. Fixed it

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherMultipleConsumers.java [hunk 3]
```diff
@@ -270,7 +270,11 @@ public synchronized void readMoreEntries() {
...
+                if (minReplayedPosition != null) {
+                    redeliveryMessages.add(minReplayedPosition.getLedgerId(), minReplayedPosition.getEntryId());
```
Comment (xrayw) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherMultipleConsumers.java:276]: Hi, Merli. I'm a little confused here, we get position by `getMessagesToReplayNow`, the returned positions not deleted in `getMessagesToReplayNow`, so why we need to add it to `redeliveryMessages` again?

**Pull request merged**